### PR TITLE
Add CVSSv4 Calculator

### PR DESCRIPTION
**Pull request by domwhewell-sage**
Working PR to add  CVSSv4 calculator onto the findings views as requested in #356.

The CVSSv4 calculator should appear as a optional tab once you click the "CVSS Calculator" accordion.
So as to minimize changes in future all Metrics are included.

### CODE REVIEW

**Comment by domwhewell-sage** [comment]
The calculator at https://www.first.org/cvss/calculator/4.0 is slightly different to previous cvss versions in that it is being distributed as a vue.js application. https://github.com/FIRSTdotorg/cvss-v4-calculator

To avoid in inaccurate calculations by converting the vue.js app to javascript and mirror the existing cvss-v3 calculator it might make sense to fork the github project into the `/static` folder and display it within an iframe (Like first.org is doing). The only custom code will be extracting the final score and vector from the iframe and placing them into the `cvss_score` and `cvss_vector` fields.

**Comment by felix-caboff** [comment]
Hi Team,

Is there anything I can do to move this particular pull request along?  Not sure what your process is, whether you need someone to do some testing or similar.  (We are desperate for it, don't worry if you have a plan, just trying to be helpful)

It looks like the CodeFactor checks failed because of the third-party library.  I would suggest that these failures are ignored as that code is not a GW specific element.

Thanks!

**Comment by chrismaddalena** [comment]
@felix-caboff Once we wrap-up some features that have been in-flight for v4.1, we'll take a look at this. The big question is how easy it is for someone to choose the calculator they want to use. Some people still want CVSS v3. Others are asking for CVSS v3.1. Another group wants CVSS v4. The ideal end state is they're all available and you can pick which one to use.

In the meantime, checking out this PR and providing feedback is very helpful.

File: ghostwriter/static/js/cvsscalc40_ui.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        else{
+          document.getElementById('id_severity').value = 1;
+        }
+    }
+}
+
+function ParseVectorCVSS4(vectorString){
+    var iframe = document.getElementById('cvsscalc40');
+    var currentSrc = iframe.src;
+    iframe.src = currentSrc.split('#')[0] + '#' + vectorString;
```
[the remark this answers is not in the document]
Comment (domwhewell-sage) [reply ghostwriter/static/js/cvsscalc40_ui.js]: EncodeURI in [2fa5754](https://github.com/GhostManager/Ghostwriter/pull/387/commits/2fa5754cad5a9e72cba398c88ed1f4639d29ab40) to escape meta characters

**Comment by felix-caboff** [comment]
@domwhewell-sage wow this looks like a lot has happened!  Thank you for prepping it for testing.  Unless you say otherwise I will try and test it this afternoon (UK time).

**Comment by felix-caboff** [comment]
For anyone reading this, please also see a competing PR #509

**Comment by felix-caboff** [comment]
Hey, you are probably best looking at the original issue that was raised: #356

**Comment by chrismaddalena** [comment]
Alex has implemented this PR along with some related feedback/requests from other issues in PR #509. I think the PRs are essentially identical in the basic functionality of offering CVSS v3 and v4 calculators. You can switch from one to the other on the fly.

I want to get this feature merged in so I'm looking to reconcile the two PRs. Is there anything missing from #509 that is in this PR?

Here is an example of the calc in PR #509:

image

**Comment by domwhewell-sage** [comment]
Hi @chrismaddalena, #509 is only including the base score, and is missing the Supplemental Metrics, Environmental (Modified Base Metrics), Environmental (Security Requirements) and Threat Metrics.
If you are the consumer those additional metrics are pretty essential to be added onto the vector, to take into account the consumers specific environment the system is deployed in.

The CVSSv3 calculator is missing the Temporal Score and Environmental Score (I know they didn't exist in the base branch anyway)

**Comment by chrismaddalena** [comment]
@domwhewell-sage We should add supplemental metrics to the calculator. That shouldn't be difficult.

**Comment by felix-caboff** [comment]
This looks shiny to me!

Not immediately important, but related....  Do the changes made in this PR open up the ability for GW to have other scoring systems in the future?  Like for example, could we do the CWSS?  (_W_eakenss not _V_ulnerability)?  It looks to me that so long as there is a scope and a vector that isn't using wildly different character sets we are good.

**Comment by chrismaddalena** [comment]
@felix-caboff Probably. The calculators are built with JavaScript, so it seems feasible to allow you to switch to more than just CVSS v3 or v4. A quick search didn't return any readily available CWSS calculators for JavaScript, so the most significant lift would probably be developing that.

**Comment by felix-caboff** [comment]
@chrismaddalena If I prep that CWSS JS, how do you want me to provide it?  I'm not in a position to implement / test integration to GW at the moment but I think I can probably provide a basic HTML / JS demonstrator in the next couple of days?

**Pull request closed without merging**